xxhr/impl/session-curl.hpp

```diff
@@ -0,0 +1,725 @@
+    prepareCommon();
+}
+
+void Session::Impl::HEAD() {
+    PrepareHead();
+    on_response(makeRequest());
+}
+
+void Session::Impl::PrepareOptions() {
+    curl_easy_setopt(curl_->handle, CURLOPT_NOBODY, 0L);
+    curl_easy_setopt(curl_->handle, CURLOPT_CUSTOMREQUEST, "OPTIONS");
+    prepareCommon();
+}
+
+void Session::Impl::OPTIONS() {
+    PrepareOptions();
+    on_response(makeRequest());
+}
+
+void Session::Impl::PreparePatch() {
+    curl_easy_setopt(curl_->handle, CURLOPT_NOBODY, 0L);
+    curl_easy_setopt(curl_->handle, CURLOPT_CUSTOMREQUEST, "PATCH");
+    prepareCommon();
+}
+
+void Session::Impl::PATCH() {
+    PreparePatch();
+    on_response(makeRequest());
+}
+
+void Session::Impl::PreparePost() {
+    curl_easy_setopt(curl_->handle, CURLOPT_NOBODY, 0L);
+
+    // In case there is no body or payload set it to an empty post:
+    if (hasBodyOrPayload_) {
+        curl_easy_setopt(curl_->handle, CURLOPT_CUSTOMREQUEST, nullptr);
+    } else {
+        curl_easy_setopt(curl_->handle, CURLOPT_POSTFIELDS, "");
+        curl_easy_setopt(curl_->handle, CURLOPT_CUSTOMREQUEST, "POST");
+    }
+    prepareCommon();
+}
+
+void Session::Impl::POST() {
+    PreparePost();
+    on_response(makeRequest());
+}
+
+void Session::Impl::PreparePut() {
+    curl_easy_setopt(curl_->handle, CURLOPT_NOBODY, 0L);
+    curl_easy_setopt(curl_->handle, CURLOPT_CUSTOMREQUEST, "PUT");
+    prepareCommon();
+}
+
+void Session::Impl::PUT() {
+    PreparePut();
+    on_response(makeRequest());
+}
+
+std::shared_ptr<CurlHolder> Session::Impl::GetCurlHolder() {
+    return curl_;
+}
+
+void Session::Impl::prepareCommon() {
+    assert(curl_->handle);
+
+    // Set Header:
+    SetHeaderInternal();
+
+    const std::string parametersContent = parameters_.content;
+    if (!parametersContent.empty()) {
+        Url new_url{url_ + "?" + parametersContent};
+        curl_easy_setopt(curl_->handle, CURLOPT_URL, new_url.c_str());
+    } else {
+        curl_easy_setopt(curl_->handle, CURLOPT_URL, url_.c_str());
+    }
+
+    // Proxy:
+    std::string protocol = url_.substr(0, url_.find(':'));
+    if (proxies_.has(protocol)) {
+        curl_easy_setopt(curl_->handle, CURLOPT_PROXY, proxies_[protocol].c_str());
+        if (proxyAuth_.has(protocol)) {
+            curl_easy_setopt(curl_->handle, CURLOPT_PROXYAUTH, CURLAUTH_ANY);
+            curl_easy_setopt(curl_->handle, CURLOPT_PROXYUSERPWD, proxyAuth_[protocol]);
+        }
+    }
+
+#if LIBCURL_VERSION_MAJOR >= 7
+#if LIBCURL_VERSION_MINOR >= 21
+    /* enable all supported built-in compressions */
+    curl_easy_setopt(curl_->handle, CURLOPT_ACCEPT_ENCODING, nullptr);
+#endif
+#endif
+
+#if LIBCURL_VERSION_MAJOR >= 7
+#if LIBCURL_VERSION_MINOR >= 71
+    // Fix loading certs from Windows cert store when using OpenSSL:
+    curl_easy_setopt(curl_->handle, CURLOPT_SSL_OPTIONS, CURLSSLOPT_NATIVE_CA);
+#endif
+#endif
+
+    curl_->error[0] = '\0';
+
+    if(!has_download_target_) {
+      response_string_.clear();
```

Comment:
Why no more clearing  in the general case ? If you have a precedent request you might carry garbage in there in the download case.